README

emlog -- the EMbedded-system LOG-device

Version 0.40, 13 August 2001

Author:   Jeremy Elson <jelson@circlemud.org>
Web page: http://www.circlemud.org/~jelson/software/emlog

--------------------------------------------------------------------------


What is emlog?
==============

emlog is a Linux kernel module that makes it easy to access the most
recent (and *only* the most recent) output from a process.  It works
just like "tail -f" on a log file, except that the storage required
never grows.  This can be useful in embedded systems where there isn't
enough memory or disk space for keeping complete log files, but the
most recent debugging messages are sometimes needed (e.g., after an
error is observed).

The emlog kernel module implements simple character device driver.
The driver acts like a named pipe that has a finite, circular buffer.
The size of the buffer is easily configurable.  As more data is
written into the buffer, the oldest data is discarded.  A process that
reads from an emlog device will first read the existing buffer, then
see new text as it's written, similar to monitoring a log file using
"tail -f".  (Non-blocking reads are also supported, if a process needs
to get the current contents of the log without blocking to wait for
new data.)

The current version of emlog should work under just about any Linux
kernel in the 2.x series, including 2.4.

emlog is free software, distributed under the GNU General Public
License (GPL); see the file COPYING (in the distribution) for details.


How is emlog used?
==================

1: Configure, compile, and install emlog

   First, decide which major number you would like to use for emlog.
   This is configured in emlog.h using the constant
   EMLOG_MAJOR_NUMBER.  The default is 241, which is in the
   "local/experimental use" range according to the kernel
   documentation (similar to the 10/8 or 192.168/16 IP networks).
   Setting the major number to 0 will cause the kernel to dynamically
   assign a major number to emlog.

   Next, if your kernel source tree is not rooted in the default
   directory of /usr/src/linux, modify the Makefile constant
   KERNEL_HOME to reflect its actual location.

   After configuration, compile by typing 'make'.  Two files should be
   generated: the kernel module itself (emlog.o), and the nbcat
   utility that will be described later.

   Finally, insert the module into the kernel using the 'insmod'
   command; e.g. 'insmod emlog.o'.  If successful, a message similar
   to "emlog: version 0.40 running, using major number 241" should
   show up in your kernel log (type 'dmesg' or 'cat /proc/kmsg' to see
   it).  You can also verify that the module has been inserted by
   typing 'cat /proc/modules'.


2: Create device files for emlog

   Next, you must use 'mknod' to create device files that your
   processes can write to.  The major number of the device files
   should be whatever number you selected in Step 1 (e.g., 241).  The
   minor number is used to indicate the *size* of the ring
   buffer for that device file, specified as the the number of
   kilobytes (e.g., 1024 bytes).  For example, to create an 8K buffer
   called 'testlog':

   % mknod /tmp/testlog c 241 8

   You can create as many devices as you like.  Internally, emlog uses
   the file's inode and device numbers to identify the buffer to which
   the file refers.


3: Write to and read from your new device file
   Once the device file has been created, simply write to your device
   file as you would any normal named pipe, e.g.

   % echo hello > /tmp/testlog

   Writes to the log will never block because the buffer never runs
   out of space; old data is simply overwritten by new data.

   You can read from the log in the normal way, e.g. using cat.  By
   default, reads block, just like "tail -f", waiting for new log
   data.  For example:

   % cat /tmp/testlog
   hello  [we immediately see the hello that we wrote in the previous step]
   _      [... and here's the cursor.  the 'cat' process is now
           blocked, waiting for new input.  New data will be displayed
           as it is written to the device by other processes.]
   ^C     [use control-c, for example, to stop reading.]

   As of version 0.40, emlog's buffers can be read and/or monitored
   by multiple concurrent readers correctly.  Data written to an
   emlog device will not disappear until it is overwritten by newer
   data, or the emlog module is removed.  (In versions 0.30 and
   earlier, data was removed from the buffer the first time it was
   read.)


4: Remove emlog when you're done

   Type 'rmmod emlog' will remove the emlog kernel module and free all
   associated buffers.  This won't work until all emlog device files
   are closed.


Other Usage Notes
=================

* emlog will allocate a fixed-size buffer on behalf of a device file
if one of the following two conditions is true:

  1-  A process has the file open for reading or writing
  2-  A process has written text to the pipe

In other words, buffers are persistent, even after a process closes
the emlog device.  Therefore, it is possible (naturally) to fill
virtual memory by creating many large emlog devices and writing one
byte to all of them.  Don't do that.  All buffers will be freed when
the emlog kernel module is removed.

* Non-blocking reads work; i.e., setting O_NONBLOCK using ioctl()
will cause an EAGAIN to be returned if there is no data ready.  In
addition, the select() and poll() functions will work correctly on
emlog devices.

* A small utility, 'nbcat', is included with the emlog distribution.
nbcat is similar to 'cat', but uses nonblocking reads.
This utility can be used to copy the current contents of an emlog
device without blocking to wait for more input.  For example:

   nbcat /var/log/emlog-device-instance > /tmp/saved-log-file

...will copy the current contents of the named emlog device to a file
in /tmp.


Emlog and devfs
===============

I love devfs and use it extensively, but I don't think it makes much
sense to use emlog with devfs.  emlog lets you create as many log
devices as you like, anywhere on the filesystem -- the module tells
them apart based on their inode number.  Having a single log device
always exist in a single place (/dev) is much less useful.  So, I have
intentionally continued using the old register_chrdev interface
instead of using the new devfs_register_chrdev interface.

 
Troubleshooting
===============

Q:  When I try to compile emlog, I get hundreds of errors related
to header files.

A: If your kernel sources are rooted anywhere other than
/usr/src/linux, make sure you change the KERNEL_HOME variable in the
Makefile to reflect their location.  If you've recently installed new
kernel sources, make sure that you've run "make config" or "make
menuconfig" in the kernel's root (e.g. /usr/src/linux).  You don't
actually have to go through the entire configuration; just make sure
that you have a /usr/include/asm and a /usr/include/linux that are
symbolic links into your kernel source tree.


Q:  When I try to insert the module using 'insmod', I get 'I/O error'.

A:  That usually means the major device number being registered by
emlog is already in use by another device driver.  Type 'cat
/proc/devices' to see a list of major device numbers that are in use.
If there is a collision, edit emlog.h and change emlog's major device
number to an unused number (or, change it to 0 in order to get a
dynamically assigned major number).


Q: I'm seeing "I/O error" at a time *other* then when the module is
inserted.

A:  Oops - you've found a bug in emlog.  Please report it.


Q:  When I try to access an emlog device file for reading or writing,
I get the error "no such device".

A:  This probably means either that the emlog kernel module is not
loaded; or, that the major number of the device file does not match
the major number that emlog registered.  To see which major number is
being used by emlog, type 'cat /proc/devices | grep emlog'.


Q:  When I try to access an emlog device file for reading or writing,
I get the error "invalid argument".

A:  The *minor* number of the emlog device file must be a number
between 1 and 128, representing the number of kilobytes (1,024 bytes)
that should be used for emlog's ring buffer.  Make sure you're
specifying a valid minor number in your 'mknod' statement.  Don't use
0.


Q:  I see "no memory" errors when I try opening new emlog files.

A:  Looks like you're out of virtual memory, sport.


Q:  When I try to remove the emlog driver ("rmmod emlog"), I get the
error "Device or resource busy".

A:  That means a process is currently using an emlog device.  You have
to wait until all processes close all emlog device files until the
driver can be removed.  Try using "lsof" to see which files are in use
by which processes.


Q:  I am trying to save a copy of the current emlog buffer to another
file, by typing "cp /tmp/emlog-test /tmp/saved-log-copy", but cp just
sits there forever.

A:  cp is blocked waiting for more data, just like 'cat' does when
used with an emlog device.  Use 'nbcat', the non-blocking cat utility
included with the emlog distribution; for example: 
   nbcat /tmp/emlog-test > /tmp/saved-log-copy


Q:  You've made my computer crash.

A:  Sorry.  If you can reproduce the problem I'll try to fix it.


Known Bugs
==========

None as of version 0.40.

Bug reports, patches, complaints, praise, and submissions of Central
Services Form 27B/6, are welcomed by the author (Jeremy Elson,
<jelson@circlemud.org>.


Version History
===============

Version 0.40 (August 13, 2001)
 - Concurrent readers and writers are now supported correctly (data is
   not consumed when it is first read, as it was in previous
   versions).
 - emlog's ring buffers now allocated using vmalloc instead of kmalloc
   to avoid locking large blocks of contiguous physical memory.
 - Added MODVERSIONS support
 - Added the 'nbcat' utility - similar to cat, but does not block at
   the end of the data.
 - Bug fix: both device number and inode number are now stored
   internally (instead of only the inode number).  This prevents the
   (unlikely) possibility that emlogs on different filesystems might
   share a single buffer.

Version 0.30 (March 1, 2001)
 - Now compiles correctly for 2.4 series kernels.
 - select() and poll() now work correctly on emlog devices.
 - Bug fix: all instances should not share one wait queue!

Version 0.20 (June 14, 2000)
 - Initial public release.


Who wrote emlog, and why?
=========================

Emlog was written by Jeremy Elson <jelson@circlemud.org> at the
University of Southern California's Information Sciences Institute as
part of the SCADDS project <http://www.isi.edu/scadds>.  SCADDS is an
embedded systems research project.  We use small PC/104-bus-based
single-board-PCs using Linux.  We wanted to save the debugging output
from certain processes, but since these things have 16MB of disk space
and 32MB of RAM, keeping complete log files was not an option.  These
tiny nodes do have serial ports running PPP, though, so it's possible
to walk over to a node with a laptop, plug in a serial cable, and then
telnet into the box.  Using emlog, we can always keep the most recent
debug messages from our processes; in case of an error, we can plug in
a debug console and see what went wrong.

This work was supported by DARPA under grant No. DABT63-99-1-0011 as
part of the SCADDS project, and was also made possible in part due to
support from Cisco Systems.

$Id: README,v 1.18 2001/08/13 23:22:35 jelson Exp $


ChangeLog
(02-09-2012) Andreas Neustifter <andreas.neustifter at gmail.com>
  - stability fixes
  - module init and remove rewritten

(31-08-2011) Andriy Stepanov <stanv at altlinux.ru>
  - fix build on 3.0.3 kernel
  - auto register /dev/emlog by udev
  
(14-08-2006) Nicu Pavel <npavel at mini-box.com>
  - replaced MODULE_PARM macro with module_param function

(12-06-2006)  Nicu Pavel <npavelat mini-box.com>
  - 2.6 kernel functions update from Darien version.
  - 2.6 Kernel Makefile

Darien Kindlund <kindlund at mitre.org>
  - Modified the emlog code to make it compatible with Linux 2.6 kernels.
  

test.sh
#!/bin/bash
echo ---- making emlog and nbcat
make
make nbcat
echo ---- inserting module
insmod emlog.ko
cat /proc/devices | grep emlog
ls -la -tr /sys/class/emlog
echo ---- testing module
mknod testlog c 250 8
echo "testing testlog."$RANDOM > testlog
./nbcat testlog
rm testlog
echo ---- removing module
rmmod emlog
echo ---- dmesg out
dmesg | tail


emlog.h

/*
 * EMLOG: the EMbedded-device LOGger
 *
 * Jeremy Elson <jelson@circlemud.org>
 * USC/Information Sciences Institute
 *
 * Modified By:
 * Andreas Neustifter <andreas.neustifter at gmail.com>
 * Andriy Stepanov <stanv at altlinux.ru>
 * Nicu Pavel <npavel at mini-box.com>
 * Darien Kindlund <kindlund at mitre.org>

 * This code is released under the GPL
 *
 * This is emlog version 0.41, released 13 August 2001, modified 21 February 2005
 * For more information see http://www.circlemud.org/~jelson/software/emlog
 *
 * $Id: emlog.h,v 1.6 2001/08/13 21:29:20 jelson Exp $
 */

#define EMLOG_MAX_SIZE       128        /* max size in kilobytes of a buffer */
#define DEVICE_NAME "emlog"

#define EMLOG_VERSION        "0.52"

/************************ Private Definitions *****************************/

struct emlog_info {
    wait_queue_head_t read_q;
    unsigned long i_ino;        /* Inode number of this emlog buffer */
    dev_t i_rdev;               /* Device number of this emlog buffer */
    char *data;                 /* The circular buffer data */
    int size;                   /* Size of the buffer pointed to by 'data' */
    int refcount;               /* Files that have this buffer open */
    int read_point;             /* Offset in circ. buffer of oldest data */
    int write_point;            /* Offset in circ. buffer of newest data */
    int offset;                 /* Byte number of read_point in the stream */
    struct emlog_info *next;
};

/* amount of data in the queue */
#define EMLOG_QLEN(einfo) ( (einfo)->write_point >= (einfo)->read_point ? \
         (einfo)->write_point - (einfo)->read_point : \
         (einfo)->size - (einfo)->read_point + (einfo)->write_point)

/* byte number of the last byte in the queue */
#define EMLOG_FIRST_EMPTY_BYTE(einfo) ((einfo)->offset + EMLOG_QLEN(einfo))

/* macro to make it more convenient to access the wait q */
#define EMLOG_READQ(einfo) (&((einfo)->read_q))


nbcat.c

/*
 * nbcat: a simple program similar to 'cat', but which uses
 * nonblocking reads.
 *
 * This utility can be used to copy the current contents of an emlog
 * device without blocking to wait for more input.  For example:
 *
 *        nbcat /var/log/emlog-device-instance > /tmp/saved-log-file
 *
 * ...will copy the current contents of the named emlog device to a
 * file in /tmp.
 *
 * This code is freeware and may be distributed without restriction.
 *
 * Jeremy Elson <jelson@circlemud.org>
 * August 11, 2001
 *
 * $Id: nbcat.c,v 1.2 2001/08/13 21:29:56 jelson Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

int main(int argc, char *argv[])
{
    int fd, retval;
    char buf[4096];

    if (argc != 2) {
        fprintf(stderr, "usage: %s <filename>\n", argv[0]);
        exit(1);
    }

    if ((fd = open(argv[1], O_RDONLY | O_NONBLOCK)) < 0) {
        perror(argv[1]);
        exit(1);
    }

    while ((retval = read(fd, buf, sizeof(buf))) > 0) {
        if (write(STDOUT_FILENO, buf, retval) < 0) {
            perror("writing to stdout");
            break;
        }
    }

    if (retval < 0 && errno != EAGAIN) {
        perror(argv[1]);
        exit(1);
    }

    return 0;
}


emlog.c

/*
 * EMLOG: the EMbedded-device LOGger
 *
 * Jeremy Elson <jelson@circlemud.org>
 * USC/Information Sciences Institute
 * 
 * Modified By:
 * Andreas Neustifter <andreas.neustifter at gmail.com>
 * Andriy Stepanov <stanv at altlinux.ru>
 * Nicu Pavel <npavel at mini-box.com>
 * Darien Kindlund <kindlund at mitre.org>
 *
 * This code is released under the GPL
 *
 * This is emlog version 0.41, released 13 August 2001, modified 25 July 2006
 * For more information see http://www.circlemud.org/~jelson/software/emlog
 *
 * $Id: emlog.c,v 1.7 2001/08/13 21:29:20 jelson Exp $
 */

#ifdef MODVERSIONS
#include <linux/modversions.h>
#endif

#include <linux/version.h>
#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
#include <linux/autoconf.h>
#else
#include <generated/autoconf.h>
#endif
#include <linux/stddef.h>
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/cdev.h>
#include <linux/device.h>
#include <linux/sched.h>
#include <linux/workqueue.h>
#include <linux/wait.h>
#include <linux/delay.h>
#include <linux/errno.h>
#include <linux/slab.h>
#include <linux/vmalloc.h>
#include <linux/types.h>
#include <linux/fs.h>
#include <linux/poll.h>

#include <asm/uaccess.h>
#include <linux/miscdevice.h>

#include "emlog.h"

struct emlog_info *emlog_info_list = NULL;
static int emlog_debug;

static dev_t emlog_dev_type = 0;
#define EMLOG_MINOR_BASE    0
#define EMLOG_MINOR_COUNT   256
static struct cdev *emlog_cdev = 0;
static struct class *emlog_class = 0;
static struct device *emlog_dev_reg;

module_param(emlog_debug, int, 0644);

#define MIN(x, y) ((x) < (y) ? (x) : y)

/* find the emlog-info structure associated with an inode.  returns a
 * pointer to the structure if found, NULL if not found */
static struct emlog_info *get_einfo(struct inode *inode)
{
    struct emlog_info *einfo;

    if (inode == NULL)
        return NULL;

    for (einfo = emlog_info_list; einfo != NULL; einfo = einfo->next)
        if (einfo->i_ino == inode->i_ino && einfo->i_rdev == inode->i_rdev)
            return einfo;

    return NULL;
}

/* create a new emlog buffer and its associated info structure.
 * returns an errno on failure, or 0 on success.  on success, the
 * pointer to the new struct is passed back using peinfo */
static int create_einfo(struct inode *inode, int minor,
                        struct emlog_info **peinfo)
{
    struct emlog_info *einfo;

    /* make sure the memory requirement is legal */
    if (minor < 1 || minor > EMLOG_MAX_SIZE)
        return -EINVAL;

    /* allocate space for our metadata and initialize it */
    if ((einfo = kmalloc(sizeof(struct emlog_info), GFP_KERNEL)) == NULL)
        goto struct_malloc_failed;

    memset(einfo, 0, sizeof(struct emlog_info));
    einfo->i_ino = inode->i_ino;
    einfo->i_rdev = inode->i_rdev;

    init_waitqueue_head(EMLOG_READQ(einfo));

    /* figure out how much of a buffer this should be and allocate the buffer */
    einfo->size = 1024 * minor;
    if ((einfo->data =
         (char *) vmalloc(sizeof(char) * einfo->size)) == NULL)
        goto data_malloc_failed;

    /* add it to our linked list */
    einfo->next = emlog_info_list;
    emlog_info_list = einfo;

    if (emlog_debug)
        printk(KERN_DEBUG "%s: allocating resources associated with inode %ld.\n", DEVICE_NAME, einfo->i_ino);

    /* pass the struct back */
    *peinfo = einfo;
    return 0;

#if 0
  other_failure:               /* if we check for other errors later, jump here */
#endif
    vfree(einfo->data);
  data_malloc_failed:
    kfree(einfo);
  struct_malloc_failed:
    return -ENOMEM;
}

/* this frees all data associated with an emlog_info buffer, including
 * the struct that you pass to the function.  don't dereference this
 * structure after calling free_einfo! */
void free_einfo(struct emlog_info *einfo)
{
    struct emlog_info **ptr;

    if (einfo == NULL) {
        printk(KERN_ERR "%s: null passed to free_einfo.\n", DEVICE_NAME);
        return;
    }

    if (emlog_debug)
        printk(KERN_DEBUG "%s: freeing resources associated with inode %ld.\n", DEVICE_NAME, einfo->i_ino);

    vfree(einfo->data);

    /* now delete the 'einfo' structure from the linked list.  'ptr' is
     * the pointer that needs to be changed... which is either the list
     * head or one of the 'next' pointers on the list. */
    ptr = &emlog_info_list;
    while (*ptr != einfo) {
        if (!*ptr) {
            printk(KERN_ERR "%s: corrupt einfo list.\n", DEVICE_NAME);
            break;
        } else
            ptr = &((**ptr).next);

    }
    *ptr = einfo->next;

}

/************************ File Interface Functions ************************/

static int emlog_open(struct inode *inode, struct file *file)
{
    int minor = MINOR(inode->i_rdev);

    struct emlog_info *einfo = NULL;
    int retval;

    if ((einfo = get_einfo(inode)) == NULL) {
        /* never heard of this inode before... create a new record */
        if ((retval = create_einfo(inode, minor, &einfo)) < 0)
            return retval;
    }

    if (einfo == NULL) {
        printk(KERN_ERR "%s: can not fetch einfo for inode %ld, in emlog_open.\n", DEVICE_NAME, inode->i_ino);
        return -EIO;
    }

    einfo->refcount++;
    try_module_get(THIS_MODULE);
    return 0;
}

/* this is called when a file is closed */
static int emlog_release(struct inode *inode, struct file *file)
{
    struct emlog_info *einfo;
    int retval = 0;

    /* get the buffer info */
    if ((einfo = get_einfo(inode)) == NULL) {
        printk(KERN_ERR "%s: can not fetch einfo for inode %ld, in emlog_release.\n", DEVICE_NAME, inode->i_ino);
        retval = EIO; goto out;
    }

    /* decrement the reference count.  if no one has this file open and
     * it's not holding any data, delete the record. */
    einfo->refcount--;

    if (einfo->refcount == 0 && EMLOG_QLEN(einfo) == 0)
        free_einfo(einfo);

  out:
    module_put(THIS_MODULE);
    return retval;
}

/* read_from_emlog reads bytes out of a circular buffer with
 * wraparound.  returns caddr_t, pointer to data read, which the
 * caller must free.  length is (a pointer to) the number of bytes to
 * be read, which will be set by this function to be the number of
 * bytes actually returned */
caddr_t read_from_emlog(struct emlog_info * einfo, size_t * length,
                        loff_t * offset)
{
    caddr_t retval;
    int bytes_copied = 0, n, start_point, remaining;

    /* is the user trying to read data that has already scrolled off? */
    if (*offset < einfo->offset)
        *offset = einfo->offset;

    /* is the user trying to read past EOF? */
    if (*offset >= EMLOG_FIRST_EMPTY_BYTE(einfo))
        return NULL;

    /* find the smaller of the total bytes we have available and what
     * the user is asking for */
    *length = MIN(*length, EMLOG_FIRST_EMPTY_BYTE(einfo) - *offset);
    remaining = *length;

    /* figure out where to start based on user's offset */
    start_point = einfo->read_point + (*offset - einfo->offset);
    start_point = start_point % einfo->size;

    /* allocate memory to return */
    if ((retval = kmalloc(sizeof(char) * remaining, GFP_KERNEL)) == NULL)
        return NULL;

    /* copy the (possibly noncontiguous) data to our buffer */
    while (remaining) {
        n = MIN(remaining, einfo->size - start_point);
        memcpy(retval + bytes_copied, einfo->data + start_point, n);
        bytes_copied += n;
        remaining -= n;
        start_point = (start_point + n) % einfo->size;
    }

    /* advance user's file pointer */
    *offset += *length;
    return retval;
}

static ssize_t emlog_read(struct file *file, char *buffer,      /* The buffer to fill with data */
                          size_t length,        /* The length of the buffer */
                          loff_t * offset)
{                               /* Our offset in the file */
    int retval;
    caddr_t data_to_return;
    struct emlog_info *einfo;

    /* get the metadata about this emlog */
    if ((einfo = get_einfo(file->f_dentry->d_inode)) == NULL) {
        printk(KERN_ERR "%s: can not fetch einfo for inode %ld, in emlog_release.\n", DEVICE_NAME, (long)(file->f_dentry->d_inode->i_ino));
        return -EIO;
    }

    /* wait until there's data available (unless we do nonblocking reads) */
    if (file->f_flags & O_NONBLOCK
        && *offset >= EMLOG_FIRST_EMPTY_BYTE(einfo))
        return -EAGAIN;

    wait_event_interruptible((einfo)->read_q,
                             *offset <
                             (einfo)->offset + EMLOG_QLEN(einfo));

    /* see if a signal woke us up */
    if (signal_pending(current))
        return -ERESTARTSYS;

    if ((data_to_return = read_from_emlog(einfo, &length, offset)) == NULL)
        return 0;

    if (copy_to_user(buffer, data_to_return, length) > 0)
        retval = -EFAULT;
    else
        retval = length;
    kfree(data_to_return);
    return retval;
}

/* write_to_emlog writes to a circular buffer with wraparound.  in the
 * case of an overflow, it overwrites the oldest unread data. */
void write_to_emlog(struct emlog_info *einfo, caddr_t buf, int length)
{
    int bytes_copied = 0;
    int overflow = 0;
    int n;

    if (length + EMLOG_QLEN(einfo) >= (einfo->size - 1)) {
        overflow = 1;

        /* in case of overflow, figure out where the new buffer will
         * begin.  we start by figuring out where the current buffer ENDS:
         * einfo->offset + EMLOG_QLEN.  we then advance the end-offset
         * by the length of the current write, and work backwards to
         * figure out what the oldest unoverwritten data will be (i.e.,
         * size of the buffer).  was that all quite clear? :-) */
        einfo->offset = einfo->offset + EMLOG_QLEN(einfo) + length
            - einfo->size + 1;
    }

    while (length) {
        /* how many contiguous bytes are available from the write point to
         * the end of the circular buffer? */
        n = MIN(length, einfo->size - einfo->write_point);
        memcpy(einfo->data + einfo->write_point, buf + bytes_copied, n);
        bytes_copied += n;
        length -= n;
        einfo->write_point = (einfo->write_point + n) % einfo->size;
    }

    /* if there is an overflow, reset the read point to read whatever is
     * the oldest data that we have, that has not yet been
     * overwritten. */
    if (overflow)
        einfo->read_point = (einfo->write_point + 1) % einfo->size;
}

static ssize_t emlog_write(struct file *file,
                           const char *buffer,
                           size_t length, loff_t * offset)
{
    caddr_t message = NULL;
    int n;
    struct emlog_info *einfo;

    /* get the metadata about this emlog */
    if ((einfo = get_einfo(file->f_dentry->d_inode)) == NULL)
        return -EIO;

    /* if the message is longer than the buffer, just take the beginning
     * of it, in hopes that the reader (if any) will have time to read
     * before we wrap around and obliterate it */
    n = MIN(length, einfo->size - 1);

    /* make sure we have the memory for it */
    if ((message = kmalloc(n, GFP_KERNEL)) == NULL)
        return -ENOMEM;

    /* copy into our temp buffer */
    if (copy_from_user(message, buffer, n) > 0) {
        kfree(message);
        return -EFAULT;
    }

    /* now copy it into the circular buffer and free our temp copy */
    write_to_emlog(einfo, message, n);
    kfree(message);

    /* wake up any readers that might be waiting for the data.  we call
     * schedule in the vague hope that a reader will run before the
     * writer's next write, to avoid losing data. */
    wake_up_interruptible(EMLOG_READQ(einfo));

    return n;
}

static unsigned int emlog_poll(struct file *file, poll_table * wait)
{
    struct emlog_info *einfo;

    /* get the metadata about this emlog */
    if ((einfo = get_einfo(file->f_dentry->d_inode)) == NULL)
        return -EIO;

    poll_wait(file, EMLOG_READQ(einfo), wait);

    if (file->f_pos < EMLOG_FIRST_EMPTY_BYTE(einfo))
        return POLLIN | POLLRDNORM;
    else
        return 0;
}

static struct file_operations emlog_fops = {
    .read = emlog_read,
    .write = emlog_write,
    .open = emlog_open,
    .release = emlog_release,
    .poll = emlog_poll,
    .llseek = no_llseek,        /* no_llseek by default introduced at v2.6.37-rc1 */
    .owner = THIS_MODULE,
};

static int __init emlog_init(void)
{
    int ret_val;

    ret_val = alloc_chrdev_region(&emlog_dev_type, EMLOG_MINOR_BASE, EMLOG_MINOR_COUNT, DEVICE_NAME);
    if(ret_val < 0) {
        printk(KERN_ERR "%s: Can not alloc_chrdev_region, error code %d.\n", DEVICE_NAME, ret_val);
        return -1;
    }

    emlog_cdev = cdev_alloc();
    if (emlog_cdev == NULL) {
        printk(KERN_ERR "%s: Can not cdev_alloc.\n", DEVICE_NAME);
        ret_val = -2; goto emlog_init_error;
    }

    emlog_cdev->ops = &emlog_fops;
    emlog_cdev->owner = THIS_MODULE;

    ret_val = cdev_add(emlog_cdev, emlog_dev_type, EMLOG_MINOR_COUNT);
    if(ret_val < 0) {
        printk(KERN_ERR "%s: Can not cdev_add, error code %d.\n", DEVICE_NAME, ret_val);
        ret_val = -3; goto emlog_init_error;
    }

    printk(KERN_INFO "%s: version %s running, major is %d, MINOR is %d.\n", DEVICE_NAME, EMLOG_VERSION, MAJOR(emlog_dev_type), MINOR(emlog_dev_type));

    emlog_class = class_create(THIS_MODULE, DEVICE_NAME);
    if (emlog_class == NULL) {
        printk(KERN_ERR "%s: Can not class_create.\n", DEVICE_NAME);
        ret_val = -4; goto emlog_init_error;
    }

    emlog_dev_reg = device_create(emlog_class, NULL, emlog_dev_type, NULL, DEVICE_NAME);
    if (emlog_dev_reg == NULL) {
        printk(KERN_ERR "%s: Can not device_create.\n", DEVICE_NAME);
        ret_val = -5; goto emlog_init_error;
    }

    goto emlog_init_okay;
  emlog_init_error:
    if (emlog_dev_reg) device_destroy(emlog_class, emlog_dev_type);
    if (emlog_class) class_destroy(emlog_class);
    if (emlog_cdev) cdev_del(emlog_cdev);
    if (emlog_dev_type) unregister_chrdev_region(emlog_dev_type, EMLOG_MINOR_COUNT);
  emlog_init_okay:
    return ret_val;
}

static void __exit emlog_remove(void)
{
    /* clean up any still-allocated memory */
    while (emlog_info_list != NULL)
        free_einfo(emlog_info_list);

    device_destroy(emlog_class, emlog_dev_type);
    class_destroy(emlog_class);
    cdev_del(emlog_cdev);
    unregister_chrdev_region(emlog_dev_type, EMLOG_MINOR_COUNT);

    printk(KERN_INFO "%s: unloaded.\n", DEVICE_NAME);
}

module_init(emlog_init);
module_exit(emlog_remove);

MODULE_LICENSE("GPL v2");

